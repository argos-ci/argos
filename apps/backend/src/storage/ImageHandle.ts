import { invariant } from "@argos/util/invariant";
import sharp from "sharp";

import type { FileHandle } from "./FileHandle";
import { tmpName } from "./tmp";

interface Dimensions {
  width: number;
  height: number;
}

export class ImageHandle {
  #fileHandle: FileHandle;
  #dimensions: Dimensions | null = null;
  #measurePromise: Promise<Dimensions> | null = null;

  constructor(args: { fileHandle: FileHandle }) {
    this.#fileHandle = args.fileHandle;
  }

  /**
   * Measure the image dimensions.
   */
  private async measure() {
    const filepath = await this.#fileHandle.getFilepath();
    const { width, height } = await sharp(filepath).metadata();
    invariant(width && height, "unable to get image dimensions");
    return { width, height };
  }

  /**
   * Get the image dimensions.
   */
  getDimensions(): Promise<Dimensions> {
    if (this.#dimensions) {
      return Promise.resolve(this.#dimensions);
    }
    if (this.#measurePromise) {
      return this.#measurePromise;
    }
    this.#measurePromise = this.measure();
    return this.#measurePromise;
  }

  /**
   * Enlarge the image to the target dimensions if needed.
   */
  async enlarge(targetDimensions: Dimensions): Promise<string> {
    const [dimensions, filepath] = await Promise.all([
      this.getDimensions(),
      this.#fileHandle.getFilepath(),
    ]);

    if (
      dimensions.width > targetDimensions.width ||
      dimensions.height > targetDimensions.height ||
      (dimensions.height === targetDimensions.height &&
        dimensions.width === targetDimensions.width)
    ) {
      return filepath;
    }

    const resultFilepath = await tmpName({ postfix: ".png" });
    await sharp(filepath)
      .resize({
        ...targetDimensions,
        fit: "contain",
        position: "left top",
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .toFile(resultFilepath);
    return resultFilepath;
  }
}
