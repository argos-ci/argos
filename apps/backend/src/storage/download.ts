import { createWriteStream } from "node:fs";
import type { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { invariant } from "@argos/util/invariant";
import type { GetObjectCommandOutput } from "@aws-sdk/client-s3";

/**
 * Download the S3 object to the specified output path.
 */
export async function download(
  result: GetObjectCommandOutput,
  outputPath: string,
) {
  invariant(result.Body, "no body");
  await pipeline(result.Body as Readable, createWriteStream(outputPath));
  return result;
}
