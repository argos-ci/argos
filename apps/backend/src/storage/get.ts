import { GetObjectCommand } from "@aws-sdk/client-s3";
import type { S3Client } from "@aws-sdk/client-s3";

export type GetParams = { s3: S3Client } & GetObjectCommand["input"];

/**
 * Get an object from S3.
 */
export async function get({ s3, ...other }: GetParams) {
  return s3.send(new GetObjectCommand(other));
}
