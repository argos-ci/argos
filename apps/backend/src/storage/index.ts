export * from "./checkIfExists";
export * from "./download";
export * from "./get";
export * from "./public-url";
export * from "./upload";
export * from "./signed-url";
export * from "./FileHandle";
export * from "./ImageHandle";
export * from "./s3";
