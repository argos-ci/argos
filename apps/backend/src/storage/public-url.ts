import config from "@/config";
import { File as FileModel } from "@/database/models/File";

import { getS3Client } from "./s3";
import { getSignedObjectUrl } from "./signed-url";

export async function getPublicUrl(key: string) {
  const s3 = getS3Client();
  const url = await getSignedObjectUrl({
    s3,
    Bucket: config.get("s3.screenshotsBucket"),
    Key: key,
    expiresIn: 3600,
    method: "GET",
  });
  return url;
}

const IMAGEKIT_PIXELS_LIMIT = 100_000_000;

function getPixelsInFile(file: FileModel) {
  if (file.isSizedImage()) {
    return file.width * file.height;
  }
  return null;
}

export function getTwicPicsUrl(key: string) {
  return new URL(key, config.get("s3.publicImageBaseUrl")).href;
}

export async function getPublicImageFileUrl(file: FileModel) {
  if (config.get("s3.publicImageBaseUrl")) {
    const pixels = getPixelsInFile(file);
    if (pixels !== null && pixels < IMAGEKIT_PIXELS_LIMIT) {
      return getTwicPicsUrl(file.key);
    }
  }

  return getPublicUrl(file.key);
}
