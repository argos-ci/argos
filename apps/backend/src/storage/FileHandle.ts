import { unlink } from "node:fs/promises";
import { promisify } from "node:util";
import { invariant } from "@argos/util/invariant";
import type { S3Client } from "@aws-sdk/client-s3";
import mime from "mime";
import { tmpName as cbTmpName } from "tmp";
import type { TmpNameCallback, TmpNameOptions } from "tmp";

import { download as s3Download } from "./download.js";
import { get as s3Get } from "./get.js";
import { ImageHandle } from "./ImageHandle.js";
import { uploadFromBuffer, uploadFromFilePath } from "./upload.js";

export const tmpName = promisify(
  (options: TmpNameOptions, cb: TmpNameCallback) => {
    cbTmpName(options, cb);
  },
);

export interface FileHandle {
  getFilepath(): Promise<string>;
  unlink(): Promise<void>;
  getImageHandle(): ImageHandle | null;
}

export class S3FileHandle implements FileHandle {
  #s3: S3Client;
  #bucket: string;
  #key: string | null;
  #filepath: string | null;
  #buffer: Buffer | null;
  #contentType: string;
  #downloadPromise: Promise<string> | null;
  #imageHandle: ImageHandle | null = null;

  constructor(args: {
    s3: S3Client;
    bucket: string;
    buffer?: Buffer;
    contentType: string;
    filepath?: string;
    key?: string;
  }) {
    this.#contentType = args.contentType;
    this.#buffer = args.buffer ?? null;
    this.#filepath = args.filepath ?? null;
    this.#s3 = args.s3;
    this.#bucket = args.bucket;
    this.#key = args.key ?? null;
    this.#downloadPromise = null;
    this.#imageHandle = args.contentType.startsWith("image/")
      ? new ImageHandle({ fileHandle: this })
      : null;
  }

  private download(): Promise<string> {
    if (this.#downloadPromise) {
      return this.#downloadPromise;
    }
    const run = async () => {
      invariant(this.#key, "missing key");
      const result = await s3Get({
        s3: this.#s3,
        Bucket: this.#bucket,
        Key: this.#key,
      });
      invariant(result.ContentType, "missing content type");
      const ext = mime.getExtension(result.ContentType);
      if (!ext) {
        throw new Error(
          `Unable to determine file extension for ${result.ContentType}`,
        );
      }
      const outputPath = await tmpName({ postfix: ext });
      await s3Download(result, outputPath);
      this.#filepath = outputPath;
      return this.#filepath;
    };
    this.#downloadPromise = run();
    return this.#downloadPromise;
  }

  getImageHandle() {
    return this.#imageHandle;
  }

  getKey() {
    invariant(this.#key, "missing key");
    return this.#key;
  }

  getFilepath() {
    if (this.#filepath) {
      return Promise.resolve(this.#filepath);
    }
    return this.download();
  }

  async unlink() {
    if (!this.#filepath && !this.#downloadPromise) {
      return;
    }
    const filepath = await this.getFilepath();
    await unlink(filepath);
  }

  async upload(): Promise<string> {
    const result = await (async () => {
      // If there is a buffer, we use it for upload
      if (this.#buffer) {
        invariant(this.#contentType, "missing content type");
        return uploadFromBuffer({
          s3: this.#s3,
          Bucket: this.#bucket,
          buffer: this.#buffer,
          contentType: this.#contentType,
          ...(this.#key && { Key: this.#key }),
        });
      }
      if (this.#filepath) {
        return uploadFromFilePath({
          s3: this.#s3,
          Bucket: this.#bucket,
          inputPath: this.#filepath,
          ...(this.#key && { Key: this.#key }),
        });
      }
      throw new Error("No filepath or buffer");
    })();

    this.#key = result.Key;
    return this.#key;
  }
}

export class LocalFileHandle implements FileHandle {
  #filepath: string;
  #imageHandle: ImageHandle | null = null;

  constructor(args: { filepath: string }) {
    this.#filepath = args.filepath;
    this.#imageHandle = mime.getType(this.#filepath)?.startsWith("image/")
      ? new ImageHandle({ fileHandle: this })
      : null;
  }

  getImageHandle() {
    return this.#imageHandle;
  }

  async getFilepath() {
    return this.#filepath;
  }

  async unlink() {
    // We do not remove images from the local filesystem (used in tests)
  }
}
