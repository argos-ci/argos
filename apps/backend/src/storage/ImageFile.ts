import { unlink } from "node:fs/promises";
import { promisify } from "node:util";
import { invariant } from "@argos/util/invariant";
import type { S3Client } from "@aws-sdk/client-s3";
import sharp from "sharp";
import { tmpName as cbTmpName } from "tmp";
import type { TmpNameCallback, TmpNameOptions } from "tmp";

import { download as s3Download } from "./download.js";
import { get as s3Get } from "./get.js";
import { uploadFromBuffer, uploadFromFilePath } from "./upload.js";

export const tmpName = promisify(
  (options: TmpNameOptions, cb: TmpNameCallback) => {
    cbTmpName(options, cb);
  },
);

interface Dimensions {
  width: number;
  height: number;
}

export interface ImageFile {
  getFilepath(): Promise<string> | string;
  getDimensions(): Promise<Dimensions>;
  unlink(): Promise<void>;
  enlarge(targetDimensions: Dimensions): Promise<string>;
}

abstract class AbstractImageFile implements ImageFile {
  abstract getFilepath(): Promise<string> | string;
  abstract unlink(): Promise<void>;

  dimensions: Dimensions | null = null;
  _measurePromise: Promise<Dimensions> | null = null;

  async measure() {
    const filepath = await this.getFilepath();
    const { width, height } = await sharp(filepath).metadata();
    invariant(width && height, "unable to get image dimensions");
    return { width, height };
  }

  async getDimensions(): Promise<Dimensions> {
    if (this.dimensions) {
      return this.dimensions;
    }
    this._measurePromise = this._measurePromise ?? this.measure();
    return this._measurePromise;
  }

  async enlarge(targetDimensions: Dimensions): Promise<string> {
    const [dimensions, filepath] = await Promise.all([
      this.getDimensions(),
      this.getFilepath(),
    ]);
    if (
      dimensions.width > targetDimensions.width ||
      dimensions.height > targetDimensions.height ||
      (dimensions.height === targetDimensions.height &&
        dimensions.width === targetDimensions.width)
    ) {
      return filepath;
    }

    const resultFilepath = await tmpName({ postfix: ".png" });
    await sharp(filepath)
      .resize({
        ...targetDimensions,
        fit: "contain",
        position: "left top",
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .toFile(resultFilepath);
    return resultFilepath;
  }
}

const getExtensionFromContentType = (contentType: string) => {
  switch (contentType) {
    case "image/png":
      return ".png";
    case "image/jpeg":
      return ".jpg";
    default:
      throw new Error(`Unsupported content type: ${contentType}`);
  }
};

export class S3ImageFile extends AbstractImageFile implements ImageFile {
  s3: S3Client;
  bucket: string;
  key: string | null;
  filepath: string | null;
  buffer: Buffer | null;
  contentType: string | null;
  downloadFromS3Promise: Promise<string> | null;

  constructor(params: {
    s3: S3Client;
    bucket: string;
    buffer?: Buffer;
    contentType?: string;
    filepath?: string;
    key?: string;
    dimensions?: Dimensions | null;
  }) {
    super();
    this.buffer = params.buffer ?? null;
    this.contentType = params.contentType ?? null;
    this.filepath = params.filepath ?? null;
    this.dimensions = params.dimensions ?? null;
    this.s3 = params.s3;
    this.bucket = params.bucket;
    this.key = params.key ?? null;
    this.downloadFromS3Promise = null;
  }

  private async downloadFromS3(): Promise<string> {
    invariant(this.key, "missing key");
    const result = await s3Get({
      s3: this.s3,
      Bucket: this.bucket,
      Key: this.key,
    });
    invariant(result.ContentType, "missing content type");
    const ext = getExtensionFromContentType(result.ContentType);
    const outputPath = await tmpName({ postfix: ext });
    await s3Download(result, outputPath);
    this.filepath = outputPath;
    return this.filepath;
  }

  async getFilepath(): Promise<string> {
    if (this.filepath) {
      return this.filepath;
    }
    this.downloadFromS3Promise =
      this.downloadFromS3Promise || this.downloadFromS3();
    return this.downloadFromS3Promise;
  }

  async unlink() {
    if (!this.filepath) {
      return;
    }
    const filepath = await this.getFilepath();
    await unlink(filepath);
  }

  async upload(): Promise<string> {
    const result = await (async () => {
      // If there is a buffer, we use it for upload
      if (this.buffer) {
        invariant(this.contentType, "missing content type");
        return uploadFromBuffer({
          s3: this.s3,
          Bucket: this.bucket,
          buffer: this.buffer,
          contentType: this.contentType,
          ...(this.key && { Key: this.key }),
        });
      }
      if (this.filepath) {
        return uploadFromFilePath({
          s3: this.s3,
          Bucket: this.bucket,
          inputPath: this.filepath,
          ...(this.key && { Key: this.key }),
        });
      }
      throw new Error("No filepath or buffer");
    })();

    this.key = result.Key;
    return this.key;
  }
}

export class LocalImageFile extends AbstractImageFile implements ImageFile {
  filepath: string;

  constructor(params: { filepath: string }) {
    super();
    this.filepath = params.filepath;
  }

  getFilepath(): string {
    return this.filepath;
  }

  async unlink() {
    // We do not remove images from the local filesystem (used in tests)
  }
}
