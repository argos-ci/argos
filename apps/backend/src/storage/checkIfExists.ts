import { get } from "./get.js";
import type { GetParams } from "./get.js";

export const checkIfExists = async ({ s3, ...other }: GetParams) => {
  try {
    await get({ s3, ...other });
    return true;
  } catch (error) {
    if (
      error instanceof Error &&
      "Code" in error &&
      error.Code === "AccessDenied"
    ) {
      return false;
    }
    throw error;
  }
};
