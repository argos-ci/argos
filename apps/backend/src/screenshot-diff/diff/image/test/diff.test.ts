import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { rimraf } from "rimraf";
import { beforeAll, describe, expect, it } from "vitest";

import { ImageHandle, LocalFileHandle } from "@/storage";

import { diffImages } from "..";
import type { DiffResult } from "../../types";

const __dirname = fileURLToPath(new URL(".", import.meta.url));

async function compareLocalImages(params: {
  baseFilepath: string;
  compareFilepath: string;
}) {
  return diffImages(
    new ImageHandle({
      fileHandle: new LocalFileHandle({
        filepath: params.baseFilepath,
      }),
    }),
    new ImageHandle({
      fileHandle: new LocalFileHandle({
        filepath: params.compareFilepath,
      }),
    }),
    { threshold: 0.5 },
  );
}

function snapshotResult(result: DiffResult) {
  return {
    score: result.score,
    width: result.file?.width,
    height: result.file?.height,
    contentType: result.file?.contentType,
  };
}

describe("diff E2E", () => {
  beforeAll(async () => {
    // Clean up actual-files
    await rimraf.sync(join(__dirname, "/actual-files"));
  });

  it("generates different images", async () => {
    const baseFilename = "checkerboard.png";
    const compareFilename = "white.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  });

  it("generates the same image", async () => {
    const baseFilename = "checkerboard.png";
    const compareFilename = "checkerboard.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(result.score).toBe(0);
  });

  it("generates different sizes images", async () => {
    const baseFilename = "checkerboard-excess.png";
    const compareFilename = "checkerboard.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  });

  it("generates the same image where 1 has a transparent background", async () => {
    const baseFilename = "checkerboard-transparent.png";
    const compareFilename = "checkerboard.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  });

  it("generates images with big diff", async () => {
    const baseFilename = "old-site.png";
    const compareFilename = "new-site.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  });

  it("works with large size image", async () => {
    const baseFilename = "big-image.png";
    const compareFilename = "big-image2.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  }, 10000);

  it("takes into account colors in comparison", async () => {
    const baseFilename = "violet-square.png";
    const compareFilename = "black-square.png";

    const result = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(snapshotResult(result)).toMatchSnapshot();
  });
});
