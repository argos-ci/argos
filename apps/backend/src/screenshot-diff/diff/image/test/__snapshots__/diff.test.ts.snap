// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`diff E2E > generates different images 1`] = `
{
  "height": 10,
  "score": 0.37,
  "width": 10,
}
`;

exports[`diff E2E > generates different sizes images 1`] = `
{
  "height": 15,
  "score": 0.5156000000000001,
  "width": 15,
}
`;

exports[`diff E2E > generates images with big diff 1`] = `
{
  "height": 4469,
  "score": 0.8096,
  "width": 1000,
}
`;

exports[`diff E2E > generates the same image where 1 has a transparent background 1`] = `
{
  "height": 10,
  "score": 0.41,
  "width": 10,
}
`;

exports[`diff E2E > takes into account colors in comparison 1`] = `
{
  "height": 10,
  "score": 1,
  "width": 10,
}
`;

exports[`diff E2E > works with large size image 1`] = `
{
  "height": 13358,
  "score": 0.2332,
  "width": 900,
}
`;
